### PR TITLE
contrib/ecl: new package (23.9.9)

### PR DESCRIPTION
**Pull request by paulapatience**

### CODE REVIEW

File: contrib/ecl/template.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+pkgname = "ecl"
+pkgver = "23.9.9"
+pkgrel = 0
+build_style = "configure"
```
Comment (nekopsykose) [inline contrib/ecl/template.py:4]: gnu_configure, and the configure.ac to reconf is present in ./src . not sure how to regen it in another folder though

File: contrib/ecl/template.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+configure_args = ["--prefix=/usr"]
```
Comment (nekopsykose) [inline contrib/ecl/template.py:5]: this is the default for gnu_configure as well

File: contrib/ecl/template.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    self.rm(self.destdir / f"usr/lib/ecl-{pkgver}/LICENSE")
+    self.rm(self.destdir / f"usr/lib/ecl-{pkgver}/build-stamp")
+
+
+@subpackage("ecl-static")
+def _static(self):
+    def func():
+        self.take(f"usr/lib/ecl-{pkgver}/*.a")
+
+    return func
```
Comment (nekopsykose) [inline contrib/ecl/template.py:36]: you can remove this as a ecl-devel-static is automatically created from the default_devel already

File: contrib/ecl/template.py (hunk 1 above)
Comment (nekopsykose) [reply contrib/ecl/template.py:4]: ah, i guess `configure_gen = ["autoreconf", "-if", "src"]`.

you'll also need `make_dir = "."` and `automake` in `hostmakedepends`

File: contrib/ecl/template.py (hunk 3 above)
Comment (nekopsykose) [reply contrib/ecl/template.py:36]: (seems you'll need that `usr/lib/ecl-{pkgver}/*.a` in the extra= of devel)

**Comment by q66** [comment]
this should have `options = ["!cross"]`, also tests are not properly enabled in this template

File: contrib/ecl/template.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+pkgdesc = "Embeddable Common Lisp"
+maintainer = "Paul A. Patience <paul@apatience.com>"
+license = "LGPL-2.1-or-later"
+url = "https://ecl.common-lisp.dev"
+source = f"https://gitlab.com/embeddable-common-lisp/{pkgname}/-/archive/{pkgver}/{pkgname}-{pkgver}.tar.gz"
+sha256 = "2ccda77461f651089d474f57d34c2fcee380ae54e734779b1bf4a4126e2a4686"
+
+
+def post_install(self):
+    self.install_license("COPYING")
```
Comment (q66) [inline contrib/ecl/template.py:23]: you don't install a license for LGPL

**Comment by nekopsykose** [comment]
seems to fail at runtime

```
$ ecl                      

;;; Unhandled lisp initialization error
;;; Message:
UNDEFINED-FUNCTION
;;; Arguments:
(:NAME EXT:SEGMENTATION-VIOLATION)
Internal or unrecoverable error in:

Lisp initialization error.

zsh: abort (core dumped)  ecl
```

**Comment by nekopsykose** [comment]
!lto fixes it. can't diagnose it more without first patching out the `libecl.so.23.9`sigsegv_handler()`, but i guess that would be ok for now

**Comment by nekopsykose** [comment]
for future reference

```
(lldb) bt
* thread #17, name = 'ecl', stop reason = signal SIGSEGV: address not mapped to object (fault address: 0x368)
  * frame #0: 0x00007fffef2be880 libecl.so.23.9`cl_parse_key(args=0x00007fffee66b900, nkey=5, keys=0x0000000000000368, vars=0x00007fffee66b8b0, rest=0x0000000000000000, allow_other_keys=0) at cmpaux.d:206:19
    frame #1: 0x00007fffef1d5825 libecl.so.23.9`L2300make_restart(narg=4) at conditions.c:190:4
    frame #2: 0x00007fffef1d5491 libecl.so.23.9`si_bind_simple_restarts(v1tag=0x00007fffef885ef1, v2names=<unavailable>) at conditions.c:455:12
    frame #3: 0x00007fffef2e2607 libecl.so.23.9`thread_entry_point(arg=0x00007fffef87fe40) at process.d:298:5
    frame #4: 0x00007fffef8b26b8 libgc.so.1`___lldb_unnamed_symbol529 + 88
```

after

```diff
--- a/src/c/unixint.d
+++ b/src/c/unixint.d
@@ -831,7 +831,7 @@
         }
 # endif /* ECL_USE_MPROTECT */
 # ifdef ECL_DOWN_STACK
-        if (sig == SIGSEGV &&
+        if (0 && sig == SIGSEGV &&
             (char*)info->si_addr > the_env->cs_barrier &&
             (char*)info->si_addr <= the_env->cs_org) {
                 unblock_signal(the_env, sig);
@@ -839,7 +839,7 @@
                 return;
         }
 # else
-        if (sig == SIGSEGV &&
+        if (0 && sig == SIGSEGV &&
             (char*)info->si_addr < the_env->cs_barrier &&
             (char*)info->si_addr >= the_env->cs_org) {
                 unblock_signal(the_env, sig);
@@ -936,7 +936,6 @@
 #endif /* !HAVE_SIGPROCMASK */
         } else if (action == ECL_T || action == @':catch') {
                 if (code == SIGSEGV) {
-                        mysignal(code, sigsegv_handler);
                 }
 #ifdef SIGBUS
                 else if (code == SIGBUS) {
```


!lto still works ok ofc

**Comment by nekopsykose** [comment]
for running tests, be inspired by https://github.com/void-linux/void-packages/commit/f29a2c78ead1240b4ad9c6727b902b023e5702d9

**Comment by paulapatience** [comment]
I've addressed all of the comments so far except running the tests, which I am looking at presently.

**Comment by paulapatience** [comment]
I've tried porting the relevant bits from [the void package][1], but the build complains that it cannot find ECL:

```shell
...
/home/paul/external/cports/bldroot/builddir/ecl-23.9.9/build/bin/ecl -norc \
                -load config.lsp \
                -eval "(ecl-tests::run-tests '())" \
                -eval "(ext:quit)" \
                2>&1 | tee check.log
/bin/sh: /home/paul/external/cports/bldroot/builddir/ecl-23.9.9/build/bin/ecl: not found
```

However, I can manually verify that the executable is there. Any ideas?

[1]: https://github.com/void-linux/void-packages/commit/f29a2c78ead1240b4ad9c6727b902b023e5702d9

**Comment by paulapatience** [comment]
Ok, with `self.chroot_cwd`, the executable is found; thanks @nekopsykose. I will now investigate the next failure.

**Comment by paulapatience** [comment]
The tests run now, but do not stop the build if they fail, which some do:

```
Did 293 tests (0 crashed), 18972 checks.
   Pass: 18953 (100%)
   Fail: 19 ( 0%)

Failure details:
--------------------------------
 MIX.0016.FIFO-TESTS:
   FAIL: (EQUAL :FOO (READ-LINE STREAM NIL :FOO))
--------------------------------
 IEEE-FP.0031.BRANCH-CUTS-SIGNED-ZERO:
   FAIL: (PLUSP (IMAGPART (ATANH Z-ABOVE)))
   FAIL: (MINUSP (IMAGPART (ATANH Z-BELOW)))
   FAIL: (PLUSP (IMAGPART (ATANH Z-ABOVE)))
   FAIL: (MINUSP (IMAGPART (ATANH Z-BELOW)))
   FAIL: (MINUSP (REALPART (ATAN Z-LEFT)))
   FAIL: (PLUSP (REALPART (ATAN Z-RIGHT)))
   FAIL: (PLUSP (IMAGPART (ATANH Z-ABOVE)))
   FAIL: (MINUSP (IMAGPART (ATANH Z-BELOW)))
   FAIL: (PLUSP (IMAGPART (ATANH Z-ABOVE)))
   FAIL: (MINUSP (IMAGPART (ATANH Z-BELOW)))
   FAIL: (MINUSP (REALPART (ATAN Z-LEFT)))
   FAIL: (PLUSP (REALPART (ATAN Z-RIGHT)))
   FAIL: (PLUSP (IMAGPART (ATANH Z-ABOVE)))
   FAIL: (MINUSP (IMAGPART (ATANH Z-BELOW)))
   FAIL: (PLUSP (IMAGPART (ATANH Z-ABOVE)))
   FAIL: (MINUSP (IMAGPART (ATANH Z-BELOW)))
   FAIL: (MINUSP (REALPART (ATAN Z-LEFT)))
   FAIL: (PLUSP (REALPART (ATAN Z-RIGHT)))
--------------------------------
```

I will look into the failures, particularly the FIFO one, later.

**Comment by paulapatience** [comment]
I've extracted the failing tests into the following script named `tests.lisp`:

```common-lisp
;;;; Run with ecl --load tests.lisp --eval "(ext:quit)"

(defmacro is (form)
  `(format t "~A ~A~%" ',form ,form))

;;; mix.0016.fifo-tests
(ext:run-program "mkfifo" '("my-fifo") :output t)
(unwind-protect
     (with-open-file (stream "my-fifo" :nonblock t)
       (with-open-file (stream2 "my-fifo" :direction :output :nonblock t)
         (write-line "foobar" stream2))
       (is (equal :foo (read-line stream nil :foo))))
  (delete-file "my-fifo"))

;;; ieee-fp.0031.branch-cuts-signed-zero
(let ((x 1.01))
  ;; branch cuts in [1,infinity)
  (let ((z-above (complex x +0.0))
        (z-below (complex x -0.0)))
    (is (plusp (imagpart (atanh z-above))))
    (is (minusp (imagpart (atanh z-below)))))
  ;; branch cuts in (-infinity,-1]
  (let ((z-above (complex (- x) +0.0))
        (z-below (complex (- x) -0.0)))
    (is (plusp (imagpart (atanh z-above))))
    (is (minusp (imagpart (atanh z-below)))))
  ;; branch cuts in [i,i*infinity)
  (let ((z-left (complex -0.0 x))
        (z-right (complex +0.0 x)))
    (is (minusp (realpart (atan z-left)))))
  ;; branch cuts in (-i*infinity,-i]
  (let ((z-left (complex -0.0 (- x)))
        (z-right (complex +0.0 (- x))))
    (is (plusp (realpart (atan z-right))))))
```

There are actually only 6 failing floating point tests, which are repeated for single, double and long floats.

**Comment by paulapatience** [comment]
I've extracted the FIFO test into a C file.

```c
#include <assert.h>
#include <fcntl.h>
#include <stdio.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

int main(void) {
  int fd, fd2;
  ssize_t nread;
  char buf[1024] = {0};

  assert(mkfifo("/tmp/my-fifo", 0666) >= 0);
  assert((fd = open("/tmp/my-fifo", O_RDONLY|O_NONBLOCK)) >= 0);
  nread = read(fd, buf, sizeof(buf));
  if (nread == 0)
    printf("EOF\n");
  else
    printf("'%s'\n", buf);

  assert((fd2 = open("/tmp/my-fifo", O_WRONLY|O_NONBLOCK)) >= 0);
  assert(write(fd2, "foobar\n", 7) == 7);
  close(fd2);

  nread = read(fd, buf, sizeof(buf));
  if (nread == 0)
    printf("EOF\n");
  else
    printf("'%s'\n", buf);
  close(fd);
  remove("/tmp/my-fifo");

  return 0;
}
```

The results are the same with musl and glibc:

```
EOF
'foobar
'
```

That's also the result of ECL with musl. ECL with glibc, on the other hand, reads EOF after closing `STREAM2`.

**Comment by paulapatience** [comment]
The tests now run and the build fails if any test does. I have disabled the failing math tests, and also the FIFO test, as it does not seem critical, particularly since the C code behaves in the same way.

File: contrib/ecl/template.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+def do_check(self):
+    build = self.chroot_cwd / "build"
+    ecl = build / "bin/ecl"
+    self.make.check(
+        args=[f"ECL={ecl}"],
+        env={
+            "LD_LIBRARY_PATH": build,
+            "TEST_IMAGE": ecl,
+            "ECLDIR": str(build) + "/",  # Trailing slash required
```
Comment (q66) [inline contrib/ecl/template.py:32]: maybe `f"{build}/"` is nicer

**Comment by q66** [comment]
i'll test this on a few other architectures before we merge it

**Comment by q66** [comment]
this PR is missing the symlinks

**Comment by q66** [comment]
test failures on aarch64: https://gist.github.com/q66/184c62bfcb629c18c3e1d45904b2b0be

on ppc64le the build/check fully passed

**Comment by paulapatience** [comment]
Added the symlinks.

I asked the maintainer about the test failures.

**Pull request merged**